pkg/cloudprovider/providers/azure/azure_loadbalancer.go

```diff
@@ -0,0 +1,610 @@
+	serviceName := getServiceName(service)
+
+	_, existsLb, err := az.getAzureLoadBalancer(lbName)
+	if err != nil {
+		return nil, false, err
+	}
+	if !existsLb {
+		glog.V(5).Infof("get(%s): lb(%s) - doesn't exist", serviceName, pipName)
+		return nil, false, nil
+	}
+
+	pip, existsPip, err := az.getPublicIPAddress(pipName)
+	if err != nil {
+		return nil, false, err
+	}
+	if !existsPip {
+		glog.V(5).Infof("get(%s): pip(%s) - doesn't exist", serviceName, pipName)
+		return nil, false, nil
+	}
+
+	return &api.LoadBalancerStatus{
+		Ingress: []api.LoadBalancerIngress{{IP: *pip.Properties.IPAddress}},
+	}, true, nil
+}
+
+// EnsureLoadBalancer creates a new load balancer 'name', or updates the existing one. Returns the status of the balancer
+func (az *Cloud) EnsureLoadBalancer(clusterName string, service *api.Service, hosts []string) (*api.LoadBalancerStatus, error) {
+	lbName := getLoadBalancerName(clusterName)
+	pipName := getPublicIPName(clusterName, service)
+	serviceName := getServiceName(service)
+	glog.V(2).Infof("ensure(%s): START clusterName=%q lbName=%q", serviceName, clusterName, lbName)
+
+	pip, err := az.ensurePublicIPExists(serviceName, pipName)
+	if err != nil {
+		return nil, err
+	}
+
+	sg, err := az.SecurityGroupsClient.Get(az.ResourceGroup, az.SecurityGroupName, "")
+	if err != nil {
+		glog.Errorf("ensure(%s): sg(%s) - retrieving failed: %q", serviceName, *sg.Name, err)
+		return nil, err
+	}
+	sg, sgNeedsUpdate, err := az.reconcileSecurityGroup(sg, clusterName, service)
+	if err != nil {
+		return nil, err
+	}
+	if sgNeedsUpdate {
+		glog.V(3).Infof("ensure(%s): sg(%s) - updating", serviceName, *sg.Name)
+		_, err := az.SecurityGroupsClient.CreateOrUpdate(az.ResourceGroup, *sg.Name, sg, nil)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	lbNeedsCreate := false
+	lb, existsLb, err := az.getAzureLoadBalancer(lbName)
+	if err != nil {
+		return nil, err
+	}
+	if !existsLb {
+		lb = network.LoadBalancer{
+			Name:       &lbName,
+			Location:   &az.Location,
+			Properties: &network.LoadBalancerPropertiesFormat{},
+		}
+		lbNeedsCreate = true
+		glog.V(3).Infof("ensure(%s): lb(%s) - needs creation")
+	}
+
+	lb, lbNeedsUpdate, err := az.reconcileLoadBalancer(lb, pip, clusterName, service, hosts)
+	if err != nil {
+		return nil, err
+	}
+	if lbNeedsCreate || lbNeedsUpdate {
+		glog.V(3).Infof("ensure(%s): lb(%s) - updating", serviceName, lbName)
+		_, err = az.LoadBalancerClient.CreateOrUpdate(az.ResourceGroup, *lb.Name, lb, nil)
+		if err != nil {
+			glog.Errorf("ensure(%s): lb(%s) - updating failed: %q", serviceName, lbName, err)
+			return nil, err
+		}
+	}
+
+	// Add the machines to the backend pool if they're not already
+	lbBackendName := getBackendPoolName(clusterName)
+	lbBackendPoolID := az.getBackendPoolID(lbName, lbBackendName)
+	hostUpdates := make([]func() error, len(hosts))
+	for i, host := range hosts {
+		f := func(serviceName, host, lbBackendPoolID string) func() error {
```

Comment:
This construction is needlessly complicated.  serviceName and lbBackendPoolID never change so you don't need to pass them around.  and if you set:
`localHost := host`

you won't need the `func(...)` indirection at all.